### PR TITLE
additional wrappers around AmazonDrive API endpoints needed for moving files

### PR DESCRIPTION
**Pull request by breunigs**

### CODE REVIEW

File: nodes.go [hunk 1]
```diff
@@ -225,6 +226,96 @@ func (n *Node) GetMetadata() (string, error) {
...
 }
 
+type updateMetadata struct {
+	Name        *string `json:"name,omitempty"`
+	Description *string `json:"description,omitempty"`
+}
+
+// Updates metadata fields name or description of Node n. If a given field is nil,
+// it will not be updated.
+func (n *Node) UpdateMetadata(name *string, description *string) (*http.Response, error) {
```
Comment (ncw) [inline nodes.go]: Can you return the Node that Amazon Drive returns please?  My policy is always to return stuff it if comes back in the API.  The caller can always ignore it...

File: nodes.go [hunk 2]
```diff
@@ -225,6 +226,96 @@ func (n *Node) GetMetadata() (string, error) {
 	return md.String(), nil
 }
 
+type updateMetadata struct {
+	Name        *string `json:"name,omitempty"`
+	Description *string `json:"description,omitempty"`
+}
+
+// Updates metadata fields name or description of Node n. If a given field is nil,
```
Comment (ncw) [inline nodes.go]: Go comments for external function should be in the style `// FuncName ...` so this should be `// UpdateMetadata changes name or ...`

https://github.com/golang/go/wiki/CodeReviewComments#comment-sentences

Though I see you removed this in the next changeset ;-)

File: nodes.go [hunk 3]
```diff
@@ -245,6 +336,20 @@ func (n *Node) Trash() (*http.Response, error) {
 
 }
 
+func (n *Node) Restore() (*http.Response, error) {
```
Comment (ncw) [inline nodes.go]: This also returns a Node which we should return to the user.

Needs a comment as it is an externally visible function.

File: nodes.go [hunk 4]
```diff
@@ -225,6 +226,96 @@ func (n *Node) GetMetadata() (string, error) {
...
+
+	err = resp.Body.Close()
+	return resp, err
+}
+
+func (n *Node) AddParent(newParentId string) (*http.Response, error) {
+	return n.changeParents(newParentId, true)
+}
+
+func (n *Node) RemoveParent(parentId string) (*http.Response, error) {
```
Comment (ncw) [inline nodes.go:265]: Needs a comment as it is an externally visible function.

File: nodes.go [hunk 5]
```diff
@@ -225,6 +226,96 @@ func (n *Node) GetMetadata() (string, error) {
...
+	if err != nil {
+		return resp, err
+	}
+	n.Parents = []string{newParentId}
+
+	err = resp.Body.Close()
+	return resp, err
+}
+
+func (n *Node) AddParent(newParentId string) (*http.Response, error) {
```
Comment (ncw) [inline nodes.go:259]: Needs a comment as it is an externally visible function.

File: nodes.go [hunk 6]
```diff
@@ -225,6 +226,96 @@ func (n *Node) GetMetadata() (string, error) {
...
+	err = resp.Body.Close()
+	return resp, err
+}
+
+type replaceParent struct {
+	FromParent string `json:"fromParent"`
+	ChildId    string `json:"childId"`
+}
+
+func (n *Node) ReplaceParent(oldParentId string, newParentId string) (*http.Response, error) {
```
Comment (ncw) [inline nodes.go:237]: I've checked the implementation - looks good!

Needs a comment as it is an externally visible function.

File: nodes.go [hunk 7]
```diff
@@ -296,23 +296,19 @@ type renameNode struct {
 }
 
 // Rename node
-func (n *Node) Rename(newName string) (*Node, *http.Response, error) {
```
Comment (ncw) [inline nodes.go:250]: I'd prefer to return the Node that Amazon Drive returns - so I don't think we need to change this, unless you know something I don't?

**Review by ncw — CHANGES_REQUESTED** [review]
Thanks for the patch - the extra methods look very useful!

I've put a few comments inline.

Can you fixup, squash and rebase onto master please?

In particular please don't send merge commits in pull requests as it makes reviewing them a lot harder and in general isn't necessary!

**Comment by breunigs** [comment]
I think I have fixed all of your input.

**Comment by ncw** [comment]
Super job thanks! I've merged that.

**Pull request merged**